### PR TITLE
Update of LDAP Driver

### PR DESCRIPTION
**Pull request by mnewnham**
- Now supports all ADOdb methods using AD Query syntax instead of SQL
- Honors **ADODB_ASSOC_CASE**
- Honors **ADODB_FETCH_MODE**
- Supports **_setConnectionParameter()_**

### CODE REVIEW

File: drivers/adodb-ldap.inc.php [hunk 1]
```diff
@@ -32,54 +32,95 @@
...
+			return null;
 
-			$this->_connectionID = @ldap_connect( $conn_info[0], $conn_info[1] );
+		
+		if (strpos($ldapServer,'ldap://') !== 0 && strpos($ldapServer,'ldaps://') !== 0) 
+		{
+			/*
+			* If ldap SSL not explicitly specified, fall back to insecure
+			*/ 
+			$ldapServer = sprintf('ldap://%s',$ldapServer);
```
Comment (dregad) [inline drivers/adodb-ldap.inc.php:83]: I don't think that block is necessary. I would just pass $ldapServer as-is to ldap_connect(), which will accept a plain host name (without :port) and return false if it's not a valid URI.

File: drivers/adodb-ldap.inc.php [hunk 2]
```diff
@@ -32,54 +32,95 @@
...
+			if ($this->debug) 
+				ADOConnection::outp($e);
+			
 			return false;
 		}
-		if( count( $LDAP_CONNECT_OPTIONS ) > 0 ) {
+
+		if( is_array($LDAP_CONNECT_OPTIONS) && count( $LDAP_CONNECT_OPTIONS ) > 0 ) 
+		{
+			/*
```
Comment (dregad) [inline drivers/adodb-ldap.inc.php:101]: Should we output a deprecation warning here ? e.g. 
```php
trigger_error('$LDAP_CONNECT_OPTIONS is deprecated, use setConnectionParameter() instead', E_USER_DEPRECATED);
```

File: drivers/adodb-ldap.inc.php [hunk 3]
```diff
@@ -150,48 +193,90 @@ function _connect( $host, $username, $password, $ldapbase)
...
 		$this->_connectionID = false;
 	}
 
-	function SelectDB($db) {
-		$this->database = $db;
+	/**
+	 * Switches the baseDN to a new value
+	 *
+	 * @param string $baseDN
+	 * @return void
```
Comment (dregad) [inline drivers/adodb-ldap.inc.php:259]: Return type does not match, should be bool / true, or remove the `return true;` statement below.

**Review by dregad — CHANGES_REQUESTED** [review]
Some preliminary comments after a quick look at the code.

I'll look in more details and run some actual tests from work (I don't have an LDAP server to test with at home).

I'll commit some whitespace, coding guidelines and PHPDoc fixes in a bit.

**Comment by dregad** [comment]
Hi @mnewnham you never responded to my review comments.

I'll commit a targeted fix for #1009 in hotfix/5.22 branch for now, and set the milestone for this improvement to 5.23.0

File: drivers/adodb-ldap.inc.php [hunk 4]
```diff
@@ -32,54 +32,95 @@
...
+	protected $version;
+	
 
-	# Options configuration information
-	var $LDAP_CONNECT_OPTIONS;
+	/*
+	* Options configuration information
+	* @deprecated 5.23
+	*/
+	public $LDAP_CONNECT_OPTIONS;
```
[the remark this answers is not in the document]
Comment (mnewnham) [reply drivers/adodb-ldap.inc.php:48]: removed, per suggestion

File: drivers/adodb-ldap.inc.php [hunk 5]
```diff
@@ -32,54 +32,95 @@
...
+	
+	/**
+	 * Connect to a AD Server
+	 *
+	 * @param string|null $ldapServer The AD Server to connect to.
+	 * @param string|null $username The username to connect as.
+	 * @param string|null $password The password to connect with.
+	 * @param string|null $ldapbase The Base DN
+	 *
+	 * @return bool|null True if connected successfully, false if connection failed, or null if the mysqli extension
```
[the remark this answers is not in the document]
Comment (mnewnham) [reply drivers/adodb-ldap.inc.php:69]: fixed

File: drivers/adodb-ldap.inc.php (hunk 1 above)
Comment (mnewnham) [reply drivers/adodb-ldap.inc.php]: This is just backwards compatibility. You've always been able to pass the hostname without the jdap:// and have it fixed

File: drivers/adodb-ldap.inc.php (hunk 2 above)
Comment (mnewnham) [reply drivers/adodb-ldap.inc.php]: changed, per suggestion

File: drivers/adodb-ldap.inc.php (hunk 3 above)
Comment (mnewnham) [reply drivers/adodb-ldap.inc.php:253]: Fixed return definition, improved docblock

**Review by dregad — APPROVED** [review]
Thanks for the fixes. I'll resolve the merge conflict and merge the PR.

**Pull request merged**